src/go/transform-sdk/sr/client.go

```diff
@@ -14,5 +14,230 @@
 
 package sr
 
+import (
+	"encoding/binary"
+	"errors"
+	"strconv"
+	"unsafe"
+
+	"github.com/redpanda-data/redpanda/src/go/transform-sdk/internal/rwbuf"
+)
+
 // schemaId is an ID of a schema registered with schema registry
 type schemaId uint32
+
+// SchemaType is an enum for the different types of schemas that can be stored in schema registry.
+type SchemaType int
+
+const (
+	TypeAvro SchemaType = iota
+	TypeProtobuf
+	TypeJSON
+)
+
+// SchemaReference is a way for a one schema to reference another. The details
+// for how referencing is done are type specific; for example, JSON objects
+// that use the key "$ref" can refer to another schema via URL. For more details
+// on references, see the following link:
+//
+//	https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#schema-references
+//	https://docs.confluent.io/platform/current/schema-registry/develop/api.html
+type Reference struct {
+	Name    string
+	Subject string
+	Version int
+}
+
+// Schema is a schema that can be registered within schema registry
+type Schema struct {
+	Schema     string
+	Type       SchemaType
+	References []Reference
+}
+
+// SchemaSubject is a schema along with the subject, version and ID of the schema
+type SubjectSchema struct {
+	Schema
+
+	Subject string
+	Version int
+	ID      int
+}
+
+// SchemaRegistryClient is a client for interacting with the schema registry within Redpanda.
+//
+// The client provides caching out of the box, which can be configured with options.
+type SchemaRegistryClient interface {
+	// LookupSchemaById looks up a schema via it's global ID.
+	LookupSchemaById(id int) (s *Schema, err error)
+	// LookupSchemaByVersion looks up a schema via a subject for a specific version.
+	//
+	// Use version -1 to get the latest version.
+	LookupSchemaByVersion(subject string, version int) (s *SubjectSchema, err error)
+	// CreateSchema creates a schema under the given subject, returning the version and ID.
+	//
+	// If an equivalent schema already exists globally, that schema ID will be reused.
+	// If an equivalent schema already exists within that subject, this will be a noop and the
+	// existing schema will be returned.
+	CreateSchema(subject string, schema Schema) (s *SubjectSchema, err error)
+}
+
+type (
+	clientOpts struct {
+		disableCaches bool
```

Comment:
Agreed, my thought process was:

* I expect people to likely call `LookupSchemaById` for every record initially, so we should cache by default
* schemas change infrequently
* changes in schema will likely also mean code will be deployed soon (because you'll want your transform to use the new data your schema just added right?), resetting the cache.
* if you want to do something custom we should just disable caches and let you write whatever fancy caching logic you want.

I've documented the unbounded cache default more and added a simple entry count based map. My hope is that is simple enough for basic usage and more complex use cases can just disable this level of caching then use whatever custom logic they want.